java/fury-format/src/test/java/org/apache/fury/format/encoder/ImplementInterfaceTest.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fury.format.encoder;
+
+import lombok.Data;
+import org.apache.fury.annotation.FuryField;
+import org.apache.fury.format.row.binary.BinaryRow;
+import org.apache.fury.memory.MemoryBuffer;
+import org.apache.fury.memory.MemoryUtils;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class ImplementInterfaceTest {
+
+  public interface InterfaceType {
+    int getF1();
+
+    String getF2();
+
+    NestedType getNested();
+
+    PoisonPill getPoison();
+  }
+
+  public interface NestedType {
+    @FuryField(nullable = false)
+    String getF3();
+  }
+
+  public static class PoisonPill {}
+
+  @Data
+  static class ImplementInterface implements InterfaceType {
+    public int f1;
+    public String f2;
+    public NestedType nested;
+    public PoisonPill poison = new PoisonPill();
+
+    public ImplementInterface(final int f1, final String f2) {
+      this.f1 = f1;
+      this.f2 = f2;
+    }
+  }
+
+  @Data
+  static class ImplementNestedType implements NestedType {
+    public String f3;
+
+    public ImplementNestedType(final String f3) {
+      this.f3 = f3;
+    }
+  }
+
+  static {
+    Encoders.registerCustomCodec(PoisonPill.class, new PoisonPillCodec());
+    Encoders.registerLazyBeanInterface(InterfaceType.class);
+    Encoders.registerLazyBeanInterface(NestedType.class);
+  }
+
+  @Test
+  public void testInterfaceTypes() {
+    final InterfaceType bean1 = new ImplementInterface(42, "42");
+    final RowEncoder<InterfaceType> encoder = Encoders.bean(InterfaceType.class);
+    final BinaryRow row = encoder.toRow(bean1);
+    final MemoryBuffer buffer = MemoryUtils.wrap(row.toBytes());
+    row.pointTo(buffer, 0, buffer.size());
+    final InterfaceType deserializedBean = encoder.fromRow(row);
```

Comment:
Is it possible to pass `InterfaceType.class` to `fromRow`? In this way, users don't have to invoke `Encoders.registerLazyBeanInterface`. This will simplify the usage.

We can cache generated `InterfaceTypeImpl` using Weak/SoftMap

Reply:
You are right, registration is not strictly required.
I was concerned about the implementation being too eager to implement all interfaces it sees, if you accidentally have something like `Supplier` or other non-data interface in your type hierarchy.

But it will probably just cause compile errors anyway, so for v1 we can just assume all interfaces referred to from another interface are implementable, and trust the user to only give us types that are sensible. Thus there is now no change to public API.

Reply:
Weak / soft map seems to not be necessary, we can still create the codec up-front, since we know all interfaces referred to from the given input interface.

Reply:
Well, I actually did not get this to work today. Either the type checks are too loose and include random things like `interface List<E>` or they are too strong and do not include the interface type I am interested in. Having a list of known safe types to implement made this easy. I can take another look tomorrow.